Mathlib/RingTheory/Ideal/Over.lean

```diff
@@ -429,4 +429,94 @@ lemma map_eq_top_iff {R S} [CommRing R] [CommRing S]
 
 end IsDomain
 
+section ideal_liesOver
+
+variable (A : Type*) [CommSemiring A] {B : Type*} [Semiring B] [Algebra A B]
+  (P : Ideal B) (p : Ideal A)
+
+/-- The ideal obtained by pulling back the ideal `P` from `B` to `A`. -/
+abbrev under : Ideal A := Ideal.comap (algebraMap A B) P
+
+theorem under_def : under A P = Ideal.comap (algebraMap A B) P := rfl
+
+instance IsPrime.under [hP : P.IsPrime] : (under A P).IsPrime :=
+  hP.comap (algebraMap A B)
+
+/-- If `P` is a maximal ideal of `B`, then the intersection of `P` and `A` is also
+  a maximal ideal. -/
+instance IsMaximal.under (A : Type*) [CommRing A] {B : Type*} [CommRing B] (P : Ideal B)
+    [P.IsMaximal] [Algebra A B] [Algebra.IsIntegral A B] : (Ideal.under A P).IsMaximal :=
+  isMaximal_comap_of_isIntegral_of_isMaximal P
+
+variable {A}
+
+/-- `P` lies over `p` if `p` is the preimage of `P` of the `algebraMap`. -/
+class LiesOver : Prop where
+  over : p = Ideal.comap (algebraMap A B) P
+
+instance over_under : P.LiesOver (Ideal.under A P) where over := rfl
+
+variable {A : Type*} [CommSemiring A] (p : Ideal A) {B : Type*} [CommSemiring B] (P : Ideal B)
+  {C : Type*} [Semiring C] (𝔓 : Ideal C) [Algebra A B] [Algebra B C] [Algebra A C]
+  [IsScalarTower A B C]
+
+theorem LiesOver.trans [hp : 𝔓.LiesOver P] [hP : P.LiesOver p] : 𝔓.LiesOver p where
+  over := by rw [hP.over, hp.over, Ideal.comap_comap, ← IsScalarTower.algebraMap_eq]
+
+theorem LiesOver.tower_bot [hp : 𝔓.LiesOver p] [hP : 𝔓.LiesOver P] : P.LiesOver p where
+  over := by rw [hp.over, hP.over, comap_comap, ← IsScalarTower.algebraMap_eq]
+
+variable (B)
+
+instance under_liesOver_of_liesOver [𝔓.LiesOver p] : (Ideal.under B 𝔓).LiesOver p :=
+  LiesOver.tower_bot p (Ideal.under B 𝔓) 𝔓
+
+theorem eq_under_under_of_liesOver [𝔓.LiesOver p] : p = Ideal.under A (Ideal.under B 𝔓) :=
```

Comment:
You should have some lemma saying `Ideal.under A (Ideal.under B 𝔓) = Ideal.under A 𝔓` and this is probably then unnecessary.